trie/trie.go

```diff
@@ -649,3 +656,82 @@ func (t *Trie) Reset() {
 func (t *Trie) Size() int {
 	return estimateSize(t.root)
 }
+
+// ReviveTrie revives the trie from the proof cache
+func (t *Trie) ReviveTrie(proof MPTProofCache) error {
+	
+	var parent node
+	var childIndex int // If parent is a fullNode, childIndex is the index of the child node
+	
+	cacheHashIndex := 0 // Keep track of the index of the cachedHash
+	nubs := proof.cacheNubs
+
+loopNubs:
+	for _, nub := range nubs {
+		key := nub.RootHexKey
+		startNode := t.root
+		parent = nil // TODO (asyukii): When RootNode is introduced, parent node will be the RootNode instead of nil
+		childIndex = -1
+		// Traverse through the trie using RootHexKey
+
+		// Loop through the key to find hash node
+		for len(key) > 0 {
+			switch n := startNode.(type) {
+			case *shortNode:
+				if len(key) < len(n.Key) || !bytes.Equal(key[:len(n.Key)], n.Key) {
+					return fmt.Errorf("key %v not found", key)
+				} else {
+					parent = n
+					startNode = n.Val
+					key = key[len(n.Key):]
+				}
+			case *fullNode:
+				startNode = n.Children[key[0]]
+				parent = n
+				childIndex = int(key[0])
+				key = key[1:]
+			default:
+				continue loopNubs
+			}
+		}
+
+		// Attach node to parent
+		if _, ok := startNode.(hashNode); ok {
+			cachedHash := proof.cacheHashes[cacheHashIndex]
```

Comment:
May leave a TODO for later check the the node if expired?

Reply:
added